onnxruntime/contrib_ops/webgpu/quantization/bitlinear.cc

```diff
@@ -0,0 +1,180 @@
+  // Each workgroup handles one row of matrix A and tile_size rows of matrix B
+  // Uses the BitLinear-specific packing where 5 ternary weights are packed per uint8
+  return WGSL_TEMPLATE_APPLY(shader, "quantization/bitlinear_multiply_small_m.wgsl.template",
+                             WGSL_TEMPLATE_PARAMETER(tile_size, tile_size_),
+                             WGSL_TEMPLATE_PARAMETER(tile_size_k_vec, tile_size_k_vec_),
+                             WGSL_TEMPLATE_PARAMETER(sub_tile_count, kSubtileCount));
+}
+
+Status BitLinear::ComputeInternal(onnxruntime::webgpu::ComputeContext& context) const {
+  const Tensor* a = context.Input(0);
+  const Tensor* b = context.Input(1);
+
+  // Validate input shapes
+  TensorShape b_shape({N_, K_});
+  MatMulComputeHelper helper;
+  ORT_RETURN_IF_ERROR(helper.Compute(a->Shape(), b_shape, false, true));
+  auto* y = context.Output(0, helper.OutputShape());
+
+  const uint32_t data_size = onnxruntime::narrow<uint32_t>(y->Shape().Size());
+  if (data_size == 0) {
+    return Status::OK();
+  }
+
+  const uint32_t M = onnxruntime::narrow<uint32_t>(helper.M());
+  const uint32_t N = onnxruntime::narrow<uint32_t>(helper.N());
+  const uint32_t K = onnxruntime::narrow<uint32_t>(helper.K());
+
+  // Validate input B shape more specifically
+  const uint32_t kQuantizationBlockSize = 20;
+  const uint32_t kWeightsPerByte = 5;
+  // When K is not divisible by kQuantizationBlockSize, weights are padded to fit kQuantizationBlockSize.
+  // During quantization of A we also pad the resulting output to K_PADDED to match the weights.
+  const uint32_t K_PADDED = ((K + (kQuantizationBlockSize - 1)) / kQuantizationBlockSize) * kQuantizationBlockSize;
+  TensorShape expected_b_shape({N, K_PADDED / kWeightsPerByte});
+  ORT_ENFORCE(b->Shape() == expected_b_shape, "Unexpected input B shape", b->Shape().ToString());
+
+  // Step 1: Quantize input A using BitLinearQuantizeProgram
+  const uint32_t quantize_output_size = (M * (K_PADDED - (K_PADDED / kWeightsPerByte)) / 4);  // skipping every 5th, packed into u32
+  const uint32_t quantize_5th_output_size = M * K_PADDED / kQuantizationBlockSize;   // every 5th element packed int u32
+
+  TensorShape quantize_output_shape({quantize_output_size});
+  TensorShape quantize_5th_output_shape({quantize_5th_output_size});
+  TensorShape scales_output_shape({M});
+
+  auto quantized_a = context.CreateGPUTensor(DataTypeImpl::GetType<uint32_t>(), quantize_output_shape);
+  auto quantized_a5 = context.CreateGPUTensor(DataTypeImpl::GetType<uint32_t>(), quantize_5th_output_shape);
+  auto scales_a = context.CreateGPUTensor(a->DataType(), scales_output_shape);
+  constexpr uint32_t kVec4Components = 4;
+  constexpr uint32_t kU32Components = 4;
+
+  {
+    BitLinearQuantizeProgram quantize_program(K, K_PADDED);
+    quantize_program
+        .SetWorkgroupSize(128)
+        .SetDispatchGroupSize(M)
```

Comment:
I am wondering for generation (M is 1), it's not efficient to only dispatch one workgroup. Maybe add another quantize shader to process vector input.